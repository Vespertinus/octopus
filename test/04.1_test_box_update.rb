class UpdateEnv < RunEnv
  def config
    super + <<EOD
object_space[0].enabled = 1
object_space[0].index[0].type = "HASH"
object_space[0].index[0].unique = 1
object_space[0].index[0].key_field[0].fieldno = 0
object_space[0].index[0].key_field[0].type = "NUM"
EOD
  end
end

env = UpdateEnv.new
env.with_server do |c1|
  c2 = env.connect_to_box
  c4 = env.connect_to_box

  wal_writer_pid = env.pid + 2 # hack!
  c1.insert [3, "baz"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t1 = Thread.new { LogPpProxy.try { c1.insert [1, "foo"] } }
  t2 = Thread.new { sleep 0.01; LogPpProxy.try { c2.insert [1, "foobar"] } }
  sleep 0.05
  c4.select 1
  Process.kill "CONT", wal_writer_pid
  puts "# wal_writer cont"
  t1.join
  t2.join
  c4.select 1
  c4.delete 1
  c4.select 1


  c1.insert [3, "bar"]
  puts "# wal_writer stop"
  Process.kill "STOP", wal_writer_pid
  t2 = Thread.new { LogPpProxy.try { c2.update_fields 3, [0, :set, 1] } }
  t1 = Thread.new { sleep 0.01; LogPpProxy.try { c1.insert [1, "foo"] } }
  sleep 0.05
  c4.select 1, 3
  Process.kill "CONT", wal_writer_pid
  puts "# wal_writer cont"
  t1.join
  t2.join
  c4.select 1

end
