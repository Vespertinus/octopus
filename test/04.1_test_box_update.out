# box.insert([3, "baz"])
1

# wal_writer stop
# box.insert([1, "foo"])
# box.insert([1, "foobar"])
Failed with: {code: 0x401, message: 'object is locked'}
# box.select(1)
[]

# wal_writer cont
1

# box.select(1)
[["\001\000\000\000", "foo"]]

# box.delete(1)
1

# box.select(1)
[]

# box.insert([3, "bar"])
2

# wal_writer stop
# box.update_fields(3, [0, :set, 1])
# box.insert([1, "foo"])
Failed with: {code: 0x401, message: 'object is locked'}
# box.select(1, 3)
[["\003\000\000\000", "bar"]]

# wal_writer cont
2

# box.select(1)
[["\001\000\000\000", "bar"]]

